README.md
ClickHouseClub
=

[![build status][travis-image]][travis-url]
[![codecov.io][codecov-image]][codecov-url]
[![David deps][david-image]][david-url]
[![node version][node-image]][node-url]

[travis-image]: https://img.shields.io/travis/cnodejs/nodeclub/master.svg?style=flat-square
[travis-url]: https://travis-ci.org/cnodejs/nodeclub
[codecov-image]: https://img.shields.io/codecov/c/github/cnodejs/nodeclub/master.svg?style=flat-square
[codecov-url]: https://codecov.io/github/cnodejs/nodeclub?branch=master
[david-image]: https://img.shields.io/david/cnodejs/nodeclub.svg?style=flat-square
[david-url]: https://david-dm.org/cnodejs/nodeclub
[node-image]: https://img.shields.io/badge/node.js-%3E=_4.2-green.svg?style=flat-square
[node-url]: http://nodejs.org/download/

## 介绍

ClickHouseClub 是使用 **Node.js** 和 **MongoDB** 开发的社区系统，界面优雅，功能丰富，小巧迅速，
已在Node.js 中文技术社区 [CNode(http://cnodejs.org)](http://cnodejs.org) & ClickHouse 中文技术社区 [ClickHouseClub(http://www.clickhouse.com.cn)](http://www.clickhouse.com.cn) 得到应用，但你完全可以用它搭建自己的社区。

## 安装部署

*不保证 Windows 系统的兼容性*

线上跑的是 [Node.js](https://nodejs.org) v4.4.0，[MongoDB](https://www.mongodb.org) 是 v3.0.5，[Redis](http://redis.io) 是 v3.0.3。

```
1. 安装 `Node.js[必须]` `MongoDB[必须]` `Redis[必须]`
2. 启动 MongoDB 和 Redis
3. `$ make install` 安装 Nodeclub 的依赖包
4. `cp config.default.js config.js` 请根据需要修改配置文件
5. `$ make build` & `$ make test` 确保各项服务都正常
6. `$ node app.js`
7. visit `http://localhost:3000`
8. done!
```

## 测试

跑测试

```bash
$ make test
```

跑覆盖率测试

```bash
$ make test-cov
```

## 贡献

有任何意见或建议都欢迎提 issue，或者直接提给 [@ifengkou](https://github.com/ifengkou)

## License

MIT


History.md
2.1.1 / 2021-03-01
==================

* add cf proxy for avater

2.1.0 / 2021-02-26
==================

* add multiavatar

2.1.0 / 2020-04-24
==================

* SEO 优化
  1. 文章内页标题应该是H1标签 (done)
  2. 链接到站外的地址，全部放上 rel=nofollow (done)
  3. 外链的文字 doom标签从 `<h3>` 改成 `<p>` (done)
  4. API页面 改为 ClickHours社区API，然后把标签改为H1 (done)
  5. 关于页面 改为 关于ClickHours社区，然后把标签改为H1 (done)
  6. 文章内容页增加面包屑导航，例如 首页 / 分享 / 【转载】趣头条基于ClickHouse玩转每天1000亿数据量 （done）

2.1.0 / 2018-10-31
==================

* 增加 color words 过滤


2.1.0 / 2015-09-15
==================

* 使用 oneapm 代替 newrelic

2.0.1 / 2015-08-07
==================

* 去掉【收藏功能】

0.3.6 / 2013-11-22
==================

  * fix #237 if topic not exists, do not modified it.
  * Merge pull request #230 from JacksonTian/fix_null
  * 修复给空值设置属性的错误
  * Merge pull request #224 from JacksonTian/config_ga
  * 将Google tracker可配置化
  * Merge pull request #223 from jiyinyiyong/ga
  * add Google Analytics
  * Merge pull request #217 from leizongmin/master
  * 编辑器进入全屏模式时，调整一下样式
  * hotfix sendAtMail
  * fixed "TypeError: Cannot read property author_id of null"
  * 使用七牛 gravatar.qiniudn.com 镜像
  * Merge pull request #212 from tjwudi/master
  * Fix 'title' textarea layout problem.
  * Merge pull request #211 from tjwudi/master
  * Add OS regular files to .gitignore
  * Merge pull request #209 from leizongmin/master
  * 解决打开Topic时自动跳到输入框问题
  * Merge pull request #205 from jiyinyiyong/master
  * Merge pull request #208 from leizongmin/master
  * 帖子内容页面，增加表格样式
  * 发布帖子使用 EpicEditor 编辑器
  * xss白名单 增加thead标签
  * Merge pull request #204 from JacksonTian/assets_issue
  * limit the length of message to 20
  * 兼容未启用压缩功能的情况
  * Merge pull request #193 from dead-horse/fix-block-count
  * fix block count
  * Merge pull request #190 from leizongmin/master
  * 修正用户收藏的话题页面，页码链接不正确问题
  * Merge pull request #186 from phoenixlzx/master
  * Fixed mongoose version
  * Merge pull request #182 from JacksonTian/assets_mini
  * 替换debug为mini
  * Merge pull request #181 from JacksonTian/assets
  * 用config.debug判断是否线上状态
  * Merge pull request #180 from JacksonTian/assets
  * 升级data2xml
  * 静态资源重构
  * Merge pull request #178 from JacksonTian/typo
  * Node.JS => Node.js
  * Merge pull request #177 from JacksonTian/style
  * 文章页面的样式
  * Fix几个样式问题
  * fixed #175 stars max size
  * fixed require package.json nae not support bug
  * Merge pull request #174 from JacksonTian/style
  * 去除用标签来制造空格的行为
  * 更新样式
  * Merge pull request #171 from jiyinyiyong/markdown-p
  * limit image height
  * limit pre-wrap inside p
  * Merge pull request #165 from VitoLau/master
  * 更新about faq页面结构
  * 调整about和faq页面 padding大小
  * Merge pull request #164 from jiyinyiyong/master
  * fix: markdown-text use pre-wrap
  * format indentations to 2
  * limit the with of message links to prevent line breaks
  * 修复文字过长没有换行的问题

0.3.5 / 2013-05-30
==================

  * Update logo based on the new official one (@finian)
  * UI more flatter; use "white-space:pre" to show spaces (@jiyinyiyong)
  * fixed #161 xss process after markdown transfer
  * 修复reply2的逻辑；暂时屏蔽标签功能
  * use fixed-width font in reply
  * a text align and a padding
  * add padding to the read messages
  * 点击回复数直接跳到最后一个回复
  * 支持html
  * fixed #154 消息跳转没有直接跳转到回复
  * fixed #146 修复tag编辑bug
  * 增加nae config
  * 增加邮件提示内容
  * read file sync package.json

0.3.4 / 2013-05-27
==================

  * user markd instead showdown, use ace (@fengmk2)
  * 使用加粗的边缘线; 过滤粉红色的边缘线; 加深 panel header 颜色; 去掉 scrollbar 定制 (@jiyinyiyong)
  * 指定xss模块的配置信息，禁止HTML标签的style和class属性 (@lezongmin)
  * shanzhai'd Github T3T (@jiyinyiyong)
  * @jiyinyiyong 修改界面布局 fixed #139
  * 添加话题详情主要内容的行高 (@kerngven)
  * 增加POST提交时间间隔限制 (@leizongmin)
  * 中英文间用空格
  * 发帖页面优化
  * 文本框高度不要闪烁
  * use bootstrap 2; hide tags
  * see the demo of new UI
  * 搜索页面，如果回复时间过长，会产生断行的情况
  * unit test cases
  * #132 Add https:// validate on user.js cnodejs/nodeclub#132 (@meteormatt)
  * Add 0.10 for travis
  * fixed #107 update user links

0.3.3 / 2013-03-11
==================

  * Merge pull request #126 from cnodejs/updateSignFlow
  * 修复topic更新bug；修复@某人 bug
  * reply2也可以定位到
  * 修复node 0.6 test cases
  * 修复删除评论异常
  * 修复 exports.updateLastReply 没有callback的bug
  * 管理员可以帮忙激活账号
  * Merge pull request #125 from JacksonTian/refine
  * Fix http to https
  * 重构注册和发帖以及发邮件的部分
  * Merge pull request #117 from JacksonTian/get_post
  * 去除掉req.method的判断，分拆方法
  * Merge pull request #122 from JacksonTian/proxy
  * 分离controller和数据操作业务逻辑
  * 改完下划线驼峰为小驼峰式风格
  * Merge pull request #116 from JacksonTian/codingstyle
  * Coding style refine.
  * Merge pull request #113 from JacksonTian/master
  * 添加依赖服务状态图标
  * Merge pull request #112 from JacksonTian/refine
  * 修正rewire.reset()导致的单元测试异常
  * Refine coding style
  * Merge pull request #111 from JacksonTian/master
  * Merge pull request #110 from JacksonTian/reset_history
  * Update Authors
  * 恢复History.md文件
  * Merge pull request #104 from ccding/master
  * fix issue #27: lower case email address for gravatar
  * Merge pull request #103 from ccding/master
  * fix issue #92: email address with gmail label ("+" encode)
  * fixed topic delete not post method security problem.
  * empty author
  * fixed author empty bug
  * Merge pull request #99 from leizongmin/master
  * 将Markdown中的H标题解析放到代码块解析后面
  * Merge pull request #96 from leizongmin/master
  * 修正无法正确解析http://127.0.0.1这样的IP地址链接
  * fixed font
  * fixed color style
  * Merge pull request #87 from jiyinyiyong/rebased
  * Merge pull request #91 from leizongmin/master
  * 使用xss模块来过滤主题及回复内容
  * update to 0.3.2
  * Merge remote branch 'cnode/master'
  * fix escape
  * Merge pull request #89 from dead-horse/master
  * fix test
  * use  in node-validator
  * Merge remote branch 'cnode/master'
  * support block code
  * Merge pull request #88 from dead-horse/master
  * fix
  * change @me to markdown
  * fix @ bug in topic content
  * Merge pull request #86 from dead-horse/master
  * not escape html in
  * add preview
  * remove tags in topics of home page
  * some css

config.default.js
/**
 * config
 */

var path = require('path');

var config = {
  // debug 为 true 时，用于本地调试
  debug: true,

  get mini_assets() { return !this.debug; }, // 是否启用静态文件的合并压缩，详见视图中的Loader

  name: 'Nodeclub', // 社区名字
  description: 'CNode：Node.js专业中文社区', // 社区的描述
  keywords: 'nodejs, node, express, connect, socket.io',

  // 添加到 html head 中的信息
  site_headers: [
    '<meta name="author" content="EDP@TAOBAO" />'
  ],
  site_logo: '/public/images/cnodejs_light.svg', // default is `name`
  site_icon: '/public/images/cnode_icon_32.png', // 默认没有 favicon, 这里填写网址
  // 右上角的导航区
  site_navs: [
    // 格式 [ path, title, [target=''] ]
    [ '/about', '关于' ]
  ],
  // cdn host，如 http://cnodejs.qiniudn.com
  site_static_host: '', // 静态文件存储域名
  // 社区的域名
  host: 'localhost',
  // 默认的Google tracker ID，自有站点请修改，申请地址：http://www.google.com/analytics/
  google_tracker_id: '',
  // 默认的cnzz tracker ID，自有站点请修改
  cnzz_tracker_id: '',

  // mongodb 配置
  db: 'mongodb://127.0.0.1/node_club_dev',

  // redis 配置，默认是本地
  redis_host: '127.0.0.1',
  redis_port: 6379,
  redis_db: 0,
  redis_password: '',

  session_secret: 'node_club_secret', // 务必修改
  auth_cookie_name: 'node_club',

  // 程序运行的端口
  port: 3000,

  // 话题列表显示的话题数量
  list_topic_count: 20,

  // RSS配置
  rss: {
    title: 'CNode：Node.js专业中文社区',
    link: 'http://cnodejs.org',
    language: 'zh-cn',
    description: 'CNode：Node.js专业中文社区',
    //最多获取的RSS Item数量
    max_rss_items: 50
  },

  log_dir: path.join(__dirname, 'logs'),

  // 邮箱配置
  mail_opts: {
    host: 'smtp.126.com',
    port: 25,
    auth: {
      user: 'club@126.com',
      pass: 'club'
    },
    ignoreTLS: true,
  },

  //weibo app key
  weibo_key: 10000000,
  weibo_id: 'your_weibo_id',

  // admin 可删除话题，编辑标签。把 user_login_name 换成你的登录名
  admins: { user_login_name: true },

  // github 登陆的配置
  GITHUB_OAUTH: {
    clientID: 'your GITHUB_CLIENT_ID',
    clientSecret: 'your GITHUB_CLIENT_SECRET',
    callbackURL: 'http://cnodejs.org/auth/github/callback'
  },
  // 是否允许直接注册（否则只能走 github 的方式）
  allow_sign_up: true,

  // oneapm 是个用来监控网站性能的服务
  oneapm_key: '',

  // 下面两个配置都是文件上传的配置

  // 7牛的access信息，用于文件上传
  qn_access: {
    accessKey: 'your access key',
    secretKey: 'your secret key',
    bucket: 'your bucket name',
    origin: 'http://your qiniu domain',
    // 如果vps在国外，请使用 http://up.qiniug.com/ ，这是七牛的国际节点
    // 如果在国内，此项请留空
    uploadURL: 'http://xxxxxxxx',
  },

  // 文件上传配置
  // 注：如果填写 qn_access，则会上传到 7牛，以下配置无效
  upload: {
    path: path.join(__dirname, 'public/upload/'),
    url: '/public/upload/'
  },

  file_limit: '1MB',

  // 版块
  tabs: [
    ['share', '分享'],
    ['ask', '问答'],
    ['job', '招聘'],
  ],
  
  //cf proxy
  avatar_cf_proxy_enable: true,
  avatar_cf_proxy_github_enable: false,
  avatar_cf_proxy_githubcontent: 'github_avatars{}.xxxx.workers.dev',
  avatar_cf_proxy_gravatar_enable: true,
  avatar_cf_proxy_gravater: 'gravatar.xxxx.workers.dev',
  
  // 极光推送
  jpush: {
    appKey: 'YourAccessKeyyyyyyyyyyyy',
    masterSecret: 'YourSecretKeyyyyyyyyyyyyy',
    isDebug: false,
  },

  create_post_per_day: 1000, // 每个用户一天可以发的主题数
  create_reply_per_day: 1000, // 每个用户一天可以发的评论数
  create_user_per_ip: 1000,
  visit_per_day: 1000, // 每个 ip 每天能访问的次数
};

if (process.env.NODE_ENV === 'test') {
  config.db = 'mongodb://127.0.0.1/node_club_test';
}

module.exports = config;
